hashicorp/terraform-provider-tfe
# Remove varset wsids computed

## Description

This PR fixes #496, where the customer was experiencing an inconsistent plan error when using `workspace_ids` as an argument that depends on value of another resource (in their example it was the workspace ids datasource). 

If you read through the issue their configuration was in fact, _correct_, and the oddity of the error was due to an inherent issue with the limitations of TF Plugin SDKv2 (and some limitations in go-tfe):

Previously, `workspace_ids` was marked as both an argument (which is the default for any attribute) and as a computed property... very commonly the schema for defining this duality sets `Optional: true` and `Computed: true`. While this is technically valid, the issue lies with how variable sets are applied to the workspace. The current implementation uses a version of go-tfe that does not support additive/subtractive behavior when applying a varset... but rather simply overwrites whatever existing variable sets have been applied with the workspace ids that are given. See [go-tfe documentation](https://pkg.go.dev/github.com/hashicorp/go-tfe#VariableSetUpdateWorkspacesOptions).

Therefore, if `workspace_ids` is set ... the attribute can never be "computed" from the API because the value overwrites whatever lives in the API. Furthermore, this is compounded when the SDK has no meaningful way to determine whether it should accept the argument's values or values computed by the API. 

## How to reproduce the error

1.  Copy pasta this configuration
```hcl
terraform {
  required_version = "~> 1.1.9"

  required_providers {
    tfe = {
      version   = "~> 0.31.0"
    }
  }
}

provider "tfe" {
  hostname = "app.terraform.io"
}

locals {
  workspace_names = ["test-cli-ws1","test-cli-ws2"]
}

resource "tfe_workspace" "test-cli-ws" {
  for_each = toset(local.workspace_names)
  name         = each.key
  organization = "foobar-inc"
  tag_names    = ["team:test-cli"]
}


data "tfe_workspace_ids" "test-cli-ws-ids" {
  names = []
  tag_names = ["team:test-cli"]
  organization = "foobar-inc"
  depends_on = [tfe_workspace.test-cli-ws]
}

resource "tfe_variable_set" "test-varset" {
  depends_on = [tfe_workspace.test-cli-ws]
  name          = "cli test Varset"
  description   = "${join(",", (keys(data.tfe_workspace_ids.test-cli-ws-ids.ids)))}"
  organization  = "foobar-inc"
  workspace_ids = values(data.tfe_workspace_ids.test-cli-ws-ids.ids)
}


output "out" {
  value = data.tfe_workspace_ids.test-cli-ws-ids.ids
}
```
2. Run `terraform init && terraform apply`. This should succeed. 
3. Now we'll append a new name to the local `workspace_names`.
```hcl
locals {
  workspace_names = ["test-cli-ws1","test-cli-ws2", "test-cli-ws3"]
}
```
4. Run `terraform apply`. We should receive the inconsistent plan error noting the change in length to `workspace_ids` in the varset resource. 

## Test the fix
1. Run `terraform destroy` to start with a clean slate. 
2. Pull these changes locally, `go install . ` and make sure you have an overrides file pointing to your local build for this provider. 
3. You're going to repeat the steps to reproduce the error, however this time the apply should succeed. Make sure to run: `TF_CLI_CONFIG_FILE=/path/to/overrides.tfrc terraform apply`
